src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -0,0 +1,138 @@
+                    // delegate void Del(int x!, int y);
+                    Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(2, 23));
+        }
+
+        [Fact]
+        public void NullCheckedAbstractMethod()
+        {
+            var source = @"
+abstract class C
+{
+    abstract public int M(int x!);
+}";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (4,31): error CS8714: Parameter 'x' can only have exclamation-point null checking in implementation methods.
+                    //     abstract public int M(int x!);
+                    Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(4, 31));
+        }
+
+        [Fact]
+        public void NullCheckedInterfaceMethod()
+        {
+            var source = @"
+interface C
+{
+    public int M(int x!);
+}";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (4,22): error CS8714: Parameter 'x' can only have exclamation-point null checking in implementation methods.
+                    //     public int M(int x!);
+                    Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(4, 22));
+        }
+
+        [ConditionalFact(typeof(MonoOrCoreClrOnly))]
+        public void NullCheckedInterfaceMethod2()
+        {
+            var source = @"
+interface C
+{
+    public void M(int x!) { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullCheckedPartialMethod()
+        {
+            var source = @"
+partial class C
+{
+    partial void M(int x!);
+}";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (4,24): error CS8714: Parameter 'x' can only have exclamation-point null checking in implementation methods.
+                    //     partial void M(int x!);
+                    Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "x").WithArguments("x").WithLocation(4, 24));
+        }
+
+        [Fact]
+        public void NullCheckedInterfaceProperty()
+        {
+            var source = @"
+interface C
+{
+    public string this[int index!] { get; set; }
+}";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (4,28): error CS8714: Parameter 'index' can only have exclamation-point null checking in implementation methods.
+                    //     public string this[int index!] { get; set; }
+                    Diagnostic(ErrorCode.ERR_MustNullCheckInImplementation, "index").WithArguments("index").WithLocation(4, 28));
+        }
+
+        [Fact]
+        public void NullCheckedAbstractProperty()
+        {
+            var source = @"
+abstract class C
+{
+    public abstract string this[int index!] { get; }
```

Comment:
Can be addressed in the next PR, but please add a test for an autoprop in a concrete class as well.

Reply:
Apparently this is not going to be a supported scenario. I disagree with that, but that's a question for LDM at a later time then.